demux/packet_pool.c

```diff
@@ -0,0 +1,100 @@
+/*
+ * This file is part of mpv.
+ *
+ * mpv is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * mpv is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "packet_pool.h"
+
+#include "common/global.h"
+#include "osdep/threads.h"
+#include "packet.h"
+
+struct demux_packet_pool {
+    mp_mutex lock;
+    struct demux_packet *packets;
+};
+
+void demux_packet_pool_init(struct mpv_global *global)
+{
+    struct demux_packet_pool *pool = talloc(global, struct demux_packet_pool);
+    talloc_set_destructor(pool, demux_packet_pool_uninit);
+    mp_mutex_init(&pool->lock);
+    pool->packets = NULL;
+
+    assert(!global->packet_pool);
+    global->packet_pool = pool;
+}
+
+void demux_packet_pool_uninit(void *p)
+{
+    struct demux_packet_pool *pool = p;
+    demux_packet_pool_clear(pool);
+    mp_mutex_destroy(&pool->lock);
+}
+
+struct demux_packet_pool *demux_packet_pool_get(struct mpv_global *global)
+{
+    // Currently all clients use the same packet pool. There is no additional
+    // state for each client, may be extended in the future.
+    return global->packet_pool;
+}
+
+void demux_packet_pool_clear(struct demux_packet_pool *pool)
+{
+    mp_mutex_lock(&pool->lock);
+    struct demux_packet *dp = pool->packets;
+    while (dp) {
+        struct demux_packet *next = dp->next;
+        free_demux_packet(dp);
+        dp = next;
+    }
+    pool->packets = NULL;
+    mp_mutex_unlock(&pool->lock);
```

Comment:
you could take the pointer under lock and then free without lock

Reply:
Sure, that works, changed. This function in fact will never be used except destroying the pool, so not much change overall.

(side note: I remember having testing async destroying of packets, but in extreme scenarious it was just stalling other thread on memory allocations, because of all the free() that it did.)